### PR TITLE
Don't enable validation when UseDefaultServiceProvider is called

### LINKED ISSUE #105134: [dotnet-sdk-9.0.100-preview.7.24366.3] Fail to launch grandnode with error "Some services are not able to be constructed"
### Description

When run the 3rd party application grandnode with the latest .NET 9 build, it failed to launch that browser displays "can't reach this page".  During the debugging, we found it failed with message "Some services are not able to be constructed" and stacktrace shows at ```Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor```. 
Because this app is very complex, and we don't identify a code snippet to reproduce the issue. So, we provided a repro machine for team to investigate. Please refer to the details from "[Internal bug](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2166741)"

### Reproduction Steps

**Repro Steps for Debug Source on repro machine:** (Please refer to the repro machine from [Internal bug ](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2166741))
The machine has dotnet-sdk-9.0.100-preview.7.24366.3 sdk and mongodb installed. 
1. Set VSDebugger_ValidateDotnetDebugLibSignatures=0 in Environment Variables.
2. Open C:\Users\test01\Desktop\Source\GrandNode\GrandNode.sln by VS 2022 preview.
3. Build Grand.Web.
4. Update Grand.Web.runtimeconfig to run against with dotnet-sdk-9.0.100-preview.7.24366.3:
```
    "frameworks": [
      {
        "name": "Microsoft.AspNetCore.App",
        "version": "9.0.0-preview.7.24365.9"
      }
    ]
```
6. Add breakpoint in Program.cs line 15.
7. Run Grand.Web in debug mode.
8. Press F10.

### Expected behavior

Loading page successfully.
image

### Actual behavior

Can't reach this page .
image
error details:
```
System.AggregateException
  HResult=0x80131500
  Message=Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: FluentValidation.IValidator`1[Grand.Web.Models.Install.InstallModel] Lifetime: Transient ImplementationType: Grand.Web.Validators.Install.InstallValidator': Unable to resolve service for type 'Grand.Services.Installation.IInstallationLocalizationService' while attempting to activate 'Grand.Web.Validators.Install.InstallValidator'.) (Error while validating the service descriptor 'ServiceType: Grand.Web.Validators.Install.InstallValidator Lifetime: Transient ImplementationType: Grand.Web.Validators.Install.InstallValidator': Unable to resolve service for type 'Grand.Services.Installation.IInstallationLocalizationService' while attempting to activate 'Grand.Web.Validators.Install.InstallValidator'.) (Error while validating the service descriptor 'ServiceType: WebEssentials.AspNetCore.Pwa.PwaController Lifetime: Transient ImplementationType: WebEssentials.AspNetCore.Pwa.PwaController': Unable to resolve service for type 'WebEssentials.AspNetCore.Pwa.PwaOptions' while attempting to activate 'WebEssentials.AspNetCore.Pwa.PwaController'.) (Error while validating the service descriptor 'ServiceType: Microsoft.AspNetCore.Mvc.Infrastructure.IActionResultExecutor`1[Microsoft.AspNetCore.Mvc.RedirectResult] Lifetime: Singleton ImplementationType: Grand.Framework.Mvc.Routing.GrandRedirectResultExecutor': Cannot consume scoped service 'Grand.Domain.Security.SecuritySettings' from singleton 'Microsoft.AspNetCore.Mvc.Infrastructure.IActionResultExecutor`1[Microsoft.AspNetCore.Mvc.RedirectResult]'.) (Error while validating the service descriptor 'ServiceType: Grand.Web.Validators.Install.InstallValidator Lifetime: Transient ImplementationType: Grand.Web.Validators.Install.InstallValidator': Unable to resolve service for type 'Grand.Services.Installation.IInstallationLocalizationService' while attempting to activate 'Grand.Web.Validators.Install.InstallValidator'.)
  Source=Microsoft.Extensions.DependencyInjection
  StackTrace:
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(ICollection`1 serviceDescriptors, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
   at Microsoft.Extensions.Hosting.HostBuilder.InitializeServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at Grand.Web.Program.Main(String[] args) in D:\grandnode\Grand.Web\Program.cs:line 15

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
InvalidOperationException: Error while validating the service descriptor 'ServiceType: FluentValidation.IValidator`1[Grand.Web.Models.Install.InstallModel] Lifetime: Transient ImplementationType: Grand.Web.Validators.Install.InstallValidator': Unable to resolve service for type 'Grand.Services.Installation.IInstallationLocalizationService' while attempting to activate 'Grand.Web.Validators.Install.InstallValidator'.

Inner Exception 2:
InvalidOperationException: Unable to resolve service for type 'Grand.Services.Installation.IInstallationLocalizationService' while attempting to activate 'Grand.Web.Validators.Install.InstallValidator'.

```

### Regression?

Yes

Verify Scenarios:
1). Windows 10 22H2 AMD64 + dotnet-sdk-3.1.426: Pass
2). Windows 10 22H2 AMD64 + dotnet-sdk-9.0.100-preview.7.24359.5: Pass
3). Windows 10 22H2 AMD64 + dotnet-sdk-9.0.100-preview.7.24366.3: Fail

### Known Workarounds

_No response_

### Configuration

**Application Name**: grandnode
**OS**： Windows 10 22H2
**CPU**: X64
**.NET Build Number**: dotnet-sdk-9.0.100-preview.7.24366.3
**App & Source , Repro machine checking at** https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2166741

**Github Link:** 
 https://github.com/grandnode/grandnode

**Dotnet Info:**
```
.NET SDK:
 Version:           9.0.100-preview.7.24366.3
 Commit:            bcf65856f2
 Workload version:  9.0.100-manifests.c564055a
 MSBuild version:   17.12.0-preview-24365-02+977d5ccf6

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\9.0.100-preview.7.24366.3\

Host:
  Version:      9.0.0-preview.7.24365.1
  Architecture: x64
  Commit:       static

.NET SDKs installed:
  9.0.100-preview.7.24366.3 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 9.0.0-preview.7.24365.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 9.0.0-preview.7.24365.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 9.0.0-preview.7.24365.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### Other information

@dotnet-actwx-bot @dotnet/compat

### PR DESCRIPTION
**Pull request by ericstj**
Fix https://github.com/dotnet/runtime/issues/105134

Not entirely sure this is the right thing, or if we should just revert the regressing change.

Please have a look @wcsanders1

### CODE REVIEW

**Comment by wcsanders1** [comment]
> Fix #105134
> 
> Not entirely sure this is the right thing, or if we should just revert the regressing change.
> 
> Please have a look @wcsanders1

It seems to me that this would correct the problem of `_defaultProviderFactoryUsed` being `true` where it should be `false`. Maybe this is not an ideal solution because future methods making assignments to `_serviceProviderFactory` would need to be aware that `_defaultProviderFactoryUsed` needs to be set. However, I don't see a better solution without more substantial refactoring.

Reverting my change would mean scope and dependency-chain validation would not happen by default in development where `Host.CreateDefaultBuilder` is not called.

**Comment by ericstj** [comment]
@wcsanders1 was this what you had in mind when you added `_defaultProviderFactoryUsed`.  Why wasn't that ever set to false in the original PR?

**Comment by wcsanders1** [comment]
> @wcsanders1 was this what you had in mind when you added `_defaultProviderFactoryUsed`. Why wasn't that ever set to false in the original PR?

This is what I had in mind. I neglected to set it to false because I failed to see that the factory method also uses that constructor.

**Comment by ericstj** [comment]
@steveharter could you review this change?

**Review by halter73 — APPROVED** [review]
Am I correct in assuming that without this fix `UseServiceProviderFactory` is completely broken in development? If so, I wonder if we should backport to preview7.

> Not entirely sure this is the right thing, or if we should just revert the regressing change.

This is better than what we had but reading the following from #97091 makes me tempted to revert this.

> It looks like this failure to validate the scope is a result of calling the `HostBuilder` constructor rather than the `CreateDefaultBuilder` method, which enables scope validation when the environment is `Development`:

I think this analysis is correct, but I don't think we should make the HostBuilder constructor behave like CreateDefaultBuilder just because we like the CreateDefaultBuilder behavior better. CreateDefaultBuilder is meant to be more opinionated.

The only reason GrandNode happened to be calling the following

```
    .UseDefaultServiceProvider((context, options) =>
    {
        options.ValidateScopes = false;
        options.ValidateOnBuild = false;
    });
```

was because it was using CreateDefaultBuilder instead of the HostBuilder constructor. I think there will be far more code that was previously using HostBuilder that won't explicitly disable validation because it wasn't necessary. While resolving a scoped service from the root container is generally a bad idea and can probably always be avoided with proper care, there's plenty of scenarios where it won't cause real problems.

**Comment by ericstj** [comment]
> If so, I wonder if we should backport to preview7.

Yes, my plan is to take this as ask for P7.

> I think there will be far more code that was previously using HostBuilder that won't explicitly disable validation because it wasn't necessary. While resolving a scoped service from the root container is generally a bad idea and can probably always be avoided with proper care, there's plenty of scenarios where it won't cause real problems.

Yes, that's why I put the breaking-change label on the original PR.  I thought the goal of that PR was to intentionally enable this in and document that as a breaking change.

Do folks think it's too breaking for the value that it provides?  We could let this go out this way in P7 (bringing this fix in) and see what folks say and revert completely it if it's too breaking.

**Comment by ericstj** [comment]
/ba-g failure here is https://github.com/dotnet/runtime/issues/105332 which is resolved as fixed with https://github.com/dotnet/runtime/pull/104516 but that was merged after this built.

**Comment by steveharter** [comment]
> Am I correct in assuming that without this fix UseServiceProviderFactory is completely broken in development? If so, I wonder if we should backport to preview7.

Yes we need this fix since `_serviceProviderFactory` will always overwrite the one assigned in UseServiceProviderFactory.

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs [hunk 1]
```diff
@@ -560,6 +560,42 @@ public void ValidateOnBuildEnabledInDevelopment()
...
+                })
+                .UseDefaultServiceProvider((context, options) =>
+                {
+                    options.ValidateScopes = false;
+                })
+                .Build();
+
+            Assert.NotNull(host.Services.GetRequiredService<ServiceA>());
+        }
+
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs:580]: Add the other case where ValidateScopes = true?
```suggestion
        [Fact]
        public void ScopeValidationtEnabledInDevelopmentWithServiceProviderChanges()
        {
            var host = new HostBuilder()
                .UseEnvironment(Environments.Development)
                .ConfigureServices(services =>
                {
                    services.AddScoped<ServiceA>();
                })
                .UseDefaultServiceProvider((context, options) =>
                {
                    options.ValidateScopes = true;
                })
                .Build();

            Assert.Throws<InvalidOperationException>(() => host.Services.GetRequiredService<ServiceA>());
        }
```

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs [hunk 2]
```diff
@@ -560,6 +560,42 @@ public void ValidateOnBuildEnabledInDevelopment()
...
+        [Fact]
+        public void ValidateOnBuildNotEnabledInDevelopmentWithServiceProviderChanges()
+        {
+            using var host = new HostBuilder()
+                .UseEnvironment(Environments.Development)
+                .ConfigureServices(serices =>
+                {
+                    serices.AddSingleton<ServiceC>();
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs:588]: So a singleton is fine here; what is this validating?

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs [hunk 3]
```diff
@@ -560,6 +560,42 @@ public void ValidateOnBuildEnabledInDevelopment()
             Assert.Throws<AggregateException>(() => hostBuilder.Build());
         }
 
+        [Fact]
+        public void ScopeValidationNotEnabledInDevelopmentWithServiceProviderChanges()
+        {
+            using var host = new HostBuilder()
+                .UseEnvironment(Environments.Development)
+                .ConfigureServices(serices =>
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs:568]: nit: "serices" -> "services" (other tests here have the same problem, so I supposed we can leave as-is)

**Review by steveharter — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs (hunk 3 above)
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs:568]: Ick, yeah, lets fix later at once.

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs (hunk 2 above)
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs:588]: I'm just inverting the tests which were added with the original PR.

Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostBuilderTests.cs:604]: Looks like `ServiceC` requires `ServiceD` so only adding `ServiceC` will fail validation.  This ensures that we don't run validation.

**Pull request merged**